core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java

```diff
@@ -0,0 +1,241 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import java.util.Map.Entry;
+
+import com.linecorp.armeria.internal.TemporaryThreadLocals;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.util.internal.StringUtil;
+
+final class QueryStringEncoder {
+
+    // Forked from netty-4.1.43.
+    // https://github.com/netty/netty/blob/bd8cea644a07890f5bada18ddff0a849b58cd861/codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java
+
+    private static final byte WRITE_UTF_UNKNOWN = (byte) '?';
+    private static final char[] CHAR_MAP = "0123456789ABCDEF".toCharArray();
+    private static final boolean[] SAFE_OCTETS =
+            createSafeOctets("-_.*abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789");
+
+    private static boolean[] createSafeOctets(String safeChars) {
+        int maxChar = -1;
+        for (int i = 0; i < safeChars.length(); i++) {
+            maxChar = Math.max(safeChars.charAt(i), maxChar);
+        }
+        final boolean[] octets = new boolean[maxChar + 1];
+        for (int i = 0; i < safeChars.length(); i++) {
+            octets[safeChars.charAt(i)] = true;
+        }
+        return octets;
+    }
+
+    static void encodeParams(TemporaryThreadLocals tempThreadLocals,
+                             StringBuilder buf, QueryParamGetters params) {
+        for (Entry<String, String> e : params) {
+            final String name = e.getKey();
+            if (isSafeOctetsOnly(name)) {
+                buf.append(name);
+            } else {
+                encodeUtf8Component(tempThreadLocals, buf, name);
+            }
+            buf.append('=');
+
+            final String value = e.getValue();
+            if (isSafeOctetsOnly(value)) {
+                buf.append(value);
+            } else {
+                encodeUtf8Component(tempThreadLocals, buf, value);
+            }
+            buf.append('&');
+        }
+
+        buf.setLength(buf.length() - 1);
+    }
+
+    private static boolean isSafeOctetsOnly(String s) {
+        final int len = s.length();
+        for (int i = 0; i < len; i++) {
+            if (!isSafeOctet(s.charAt(i))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private static boolean isSafeOctet(char c) {
+        return c < SAFE_OCTETS.length && SAFE_OCTETS[c];
+    }
+
+    /**
+     * Encodes a query component (name or value).
+     *
+     * @see ByteBufUtil#writeUtf8(ByteBuf, CharSequence, int, int)
+     */
+    private static void encodeUtf8Component(TemporaryThreadLocals tempThreadLocals,
+                                            StringBuilder buf, String s) {
+        final char[] tmp = tempThreadLocals.charArray(12);
```

Comment:
Probably better to allocate a normal array, it's small enough that it should be on the stack. Though the escape analysis is hit-or-miss here - if you allocate in each method it's almost guaranteed to work though and is similar to what guava's doing. Also, since this array has dynamic size (as opposed to `new char[12]`), Java can't remove range checks which hurts performance.

Reply:
Tried already and it seems like escape analysis does not kick in sadly.

Reply:
Escape analysis works as expected after manually inlining `writeUtf8Surrogate()` :tada: